File: e2e/clients/go/catalog_network.go
package client

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"
)

// catalogTestnetCaip2 reads testnet.caip2 from e2e/config/mechanisms_<id>.json.
func catalogTestnetCaip2(networkID string) (string, error) {
	candidates := []string{}
	if injected := os.Getenv("E2E_MECHANISMS_CATALOG"); injected != "" {
		candidates = append(candidates, injected)
	}
	dir, err := os.Getwd()
	if err != nil {
		return "", err
	}
	for {
		candidates = append(candidates, filepath.Join(dir, "config"))
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	for _, catalogDir := range candidates {
		path := filepath.Join(catalogDir, fmt.Sprintf("mechanisms_%s.json", networkID))
		raw, readErr := os.ReadFile(path)
		if readErr != nil {
			continue
		}
		var file struct {
			Testnet struct {
				Caip2 string `json:"caip2"`
			} `json:"testnet"`
		}
		if err := json.Unmarshal(raw, &file); err != nil {
			return "", fmt.Errorf("%s: %w", path, err)
		}
		if file.Testnet.Caip2 == "" {
			return "", fmt.Errorf("%s: missing testnet.caip2", path)
		}
		return file.Testnet.Caip2, nil
	}
	return "", fmt.Errorf("could not locate mechanisms_%s.json", networkID)
}

// resolveNetworkCaip2 prefers `${ID}_NETWORK`, else catalog testnet CAIP-2.
func resolveNetworkCaip2(networkID string) string {
	if fromEnv := os.Getenv(strings.ToUpper(networkID) + "_NETWORK"); fromEnv != "" {
		return fromEnv
	}
	caip2, err := catalogTestnetCaip2(networkID)
	if err != nil {
		log.Fatalf("❌ %v", err)
	}
	return caip2
}

// caip2Pattern derives a CAIP-2 namespace wildcard (eip155:*) from a concrete CAIP-2 id.
func caip2Pattern(caip2 string) string {
	ns, _, ok := strings.Cut(caip2, ":")
	if !ok || ns == "" {
		log.Fatalf("❌ invalid caip2: %s", caip2)
	}
	return ns + ":*"
}

// networkCaip2Pattern is the client registration pattern for a catalog network.
func networkCaip2Pattern(networkID string) string {
	return caip2Pattern(resolveNetworkCaip2(networkID))
}

File: e2e/clients/python/catalog_network.py
"""Catalog CAIP-2 helpers for Python e2e clients."""

from __future__ import annotations

import json
import os
from pathlib import Path


def _catalog_testnet_caip2(network_id: str) -> str:
    """Read testnet.caip2 from e2e/config/mechanisms_<id>.json."""
    injected = os.getenv("E2E_MECHANISMS_CATALOG")
    candidates: list[Path] = []
    if injected:
        candidates.append(Path(injected))
    here = Path(__file__).resolve()
    candidates.extend(parent / "config" for parent in here.parents)
    for catalog_dir in candidates:
        path = catalog_dir / f"mechanisms_{network_id}.json"
        if path.is_file():
            data = json.loads(path.read_text(encoding="utf-8"))
            return data["testnet"]["caip2"]
    raise FileNotFoundError(f"Could not locate mechanisms_{network_id}.json")


def resolve_network_caip2(network_id: str) -> str:
    """Prefer `${ID}_NETWORK`, else catalog testnet CAIP-2."""
    return os.environ.get(f"{network_id.upper()}_NETWORK") or _catalog_testnet_caip2(network_id)


def caip2_pattern(caip2: str) -> str:
    """Derive a CAIP-2 namespace wildcard (`eip155:*`) from a concrete CAIP-2 id."""
    ns = caip2.split(":", 1)[0]
    if not ns:
        raise ValueError(f"invalid caip2: {caip2}")
    return f"{ns}:*"


def network_caip2_pattern(network_id: str) -> str:
    """Client registration pattern for a catalog network id."""
    return caip2_pattern(resolve_network_caip2(network_id))
